File: Java/src/brutefarce/Bruteforcer.java
package brutefarce;

import java.util.concurrent.ThreadLocalRandom;

public class Bruteforcer {
    static final String ALPHANUMERIC = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
    private String password;
    private int minLength;
    private int maxLength;
    private long guesses = 1;
    private boolean isNotCracked = true;

    public Bruteforcer(String pass, int min, int max) {
        password = pass;
        minLength = min;
        maxLength = max;
    }

    private int getGuessLength() {
        return ThreadLocalRandom.current().nextInt(minLength, maxLength + 1);
    }

    private String newGuess(int guessLength) {
        StringBuilder stringBuilder = new StringBuilder(guessLength);

        for (int i = 0; i < guessLength; i++) {
            char randomCharacter = ALPHANUMERIC.charAt(ThreadLocalRandom.current().nextInt(0, ALPHANUMERIC.length()));
            stringBuilder.append(randomCharacter);
        }

        return stringBuilder.toString();
    }

    public long breakPassword() {
        while (isNotCracked) {
            String currentGuess = newGuess(getGuessLength());

            System.out.println(guesses + ". Guess is " + currentGuess);

            if (currentGuess.equals(password)) {
                isNotCracked = false;
                System.out.println("Found a match! " + currentGuess);
            }

            guesses++;
        }

        return guesses;
    }
}

File: Scala/src/brutefarce/Bruteforcer.scala
package brutefarce

import java.util.concurrent.ThreadLocalRandom

class Bruteforcer(password: String, minLength: Int, maxLength: Int) {
  val ALPHANUMERIC = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890"
  private var guesses: Long = 1
  private var isNotCracked = true

  private def getGuessLength(): Int = ThreadLocalRandom.current().nextInt(minLength, maxLength + 1)

  private def newGuess(guessLength: Int): String = {
    val stringBuilder = new StringBuilder(guessLength)

    for (i <- 0 until guessLength) {
      val randomCharacter = ALPHANUMERIC.charAt(ThreadLocalRandom.current().nextInt(0, ALPHANUMERIC.length))
      stringBuilder.append(randomCharacter)
    }

    stringBuilder.toString()
  }

  def breakPassword(): Long = {
    while (isNotCracked) {
      val currentGuess = newGuess(getGuessLength())

      println(s"$guesses. Guess is $currentGuess")

      if (currentGuess.equals(password)) {
        isNotCracked = false
        println(s"Found a match! $currentGuess")
      }

      guesses += 1
    }

    guesses
  }
}